mongo/integration/mock_find_test.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) MongoDB, Inc. 2021-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package integration
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// finder is an object that implements FindOne and Find.
+type finder interface {
+	FindOne(ctx context.Context, filter interface{}, opts ...*options.FindOneOptions) *mongo.SingleResult
+	Find(context.Context, interface{}, ...*options.FindOptions) (*mongo.Cursor, error)
+}
+
+// mockFinder implements finder.
+type mockFinder struct {
+	docs     []interface{}
+	err      error
+	registry *bsoncodec.Registry
+}
+
+// FindOne mocks a findOne operation using NewMockSingleResult.
+func (mf *mockFinder) FindOne(_ context.Context, _ interface{}, _ ...*options.FindOneOptions) *mongo.SingleResult {
+	res, err := mongo.NewMockSingleResult(mf.docs, mf.err, mf.registry)
+
+	// If there was an error in SingleResult creation, override the error of SingleResult.
```

Comment:
Mocking `FindOne` is slightly difficult here, as `NewMockSingleResult` _can_ return an error. I've chosen to override the `SingleResult.err` with the creation error if it exists, but ideally users would have access to both the original `mf.err` and the creation error.

Reply:
Should L40 be `res.err = err`?

Reply:
Ah great catch. You can't actually access `res.err` since it's an unexported field. I had to add some logic to `NewSingleResultFromDocument` to store potential creation errors on the returned `SingleResult` (noted this behavior in the documentation, too). Let me know what you think.